### ISSUE TITLE
Add a way to get the tree size of a jaxpr

### ISSUE THREAD

**Issue by carlosgmartin** [jax-ml/jax#24995]
**Feature request:** Add a way to get the size of a [`Jaxpr`](https://jax.readthedocs.io/en/latest/_autosummary/jax.core.Jaxpr.html) and [`ClosedJaxpr`](https://jax.readthedocs.io/en/latest/_autosummary/jax.core.ClosedJaxpr.html), in terms of the size of the tree itself (not the length of its string representation).

This could be done by adding a `__len__` dunder to these classes, or another method.

**Comment by superbobry** [comment]
Can you add a few details about what you are trying to do, please?

**Comment by jakevdp** [comment]
Does `len(jaxpr.eqns)` do what you want?

**Comment by jakevdp** [comment]
> @jakevdp No.

Maybe if you disable JIT it will give you something closer to what you're after?
```python
from jax import numpy as jnp
args = jnp.zeros([2, 3, 5]), jnp.zeros([2, 7, 5], int), 1., 1, False
with jax.disable_jit():
  jaxpr = jax.make_jaxpr(jnp.put_along_axis, [3, 4])(*args)
print(len(jaxpr.eqns))  # 24
```

**Comment by jakevdp** [comment]
Side note though: jaxpr size is not a very good proxy for complexity in general: for example, jaxprs do not do any dead code elimination. Given this, compiler-based measures like cost analysis will probably be more robust.

**Comment by carlosgmartin** [comment]
@jakevdp That's closer to what I'm looking for!

The reason I brought this up is because of [your comment](https://github.com/jax-ml/jax/pull/24871#issuecomment-2483938641):

> I think we should optimize for a simple jaxpr here, not optimize for runtime on a particular device (which may or may not reflect runtime on other device types, or with other array sizes).

which could also be good advice for other functions beyond `put_along_axis`.

**Comment by jakevdp** [comment]
Oh OK – I don't think "simple jaxpr" is necessarily the right metric in general, but in that specific case we were talking about implementing `put_along_axis, which theoretically corresponds to binding a single `scatter` call. In that particular case, we should aim to write that single `scatter` call.
